streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -754,6 +764,27 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
         }
     }
 
+  /**
+   * Maps each element of this stream to another stream and returns the
+   * non-deterministic merge of those streams, executing all inner streams
+   * concurrently.
+   * A semaphore with n permits is provided to limit the number of
+   * concurrent executions.
+   * Up to `outputBuffer` elements of the produced streams may be
+   * buffered in memory by this operator.
+   */
+  final def flatMapParSema[R1 <: R, E1 >: E, B](
```

Comment:
Not always. I found a use case for this while testing the group* combinators.
There I wanted to have a limit on the number of concurrent executions, but not on the number of groups that are actively consuming in order to avoid deadlocks. 
With this you can write something like this and get the desired behavior.
ZStream#withPermit would be too weak to do this as this requires  some 'external' semaphore.

I added ZStream#withSemaphore that makes it trivial to do this adhoc though, so maybe we should still drop this.
```scala
        .groupByKey(identity, 8192)
        .flatMapParSema(10, 16) {
          case (sema, (k, s)) =>
            s.withPermit(sema).transduce(Sink.foldLeft[String, Int](0) { case (acc: Int, _: String) => acc + 1 }).take(1).map((k -> _))
        }
```

Reply:
Not sold on either - they're just special cases for `flatMap`. I suggest we drop them.